### PR TITLE
Implement support for multiple simultaneous sessions

### PR DESCRIPTION
**Pull request by pontus**
### Description

Add support for multiusers (multiple simultaneous sessions).

One could consider doing this with a library to provide sessions, but as a non-core author, I aim to keep this PR as minimal as possible.

While I believe these changes to be correct, I do note that there it touches a lot of lines in important areas, so a bit of a warning that it might be recommend to take some time for review. 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (this needs a follow up PR)

### Changes Made

* Make sessions references indirect (accessible via session id,taken from cookie).
* Change

### Testing

- [x] Unit Tests
- [ ] Integration Tests
- [ ] Tests do not apply
- [ ] Needs testing (start an issue or do a follow up PR about it)

### CODE REVIEW

**Comment by blankdots** [comment]
a good library would be https://github.com/aio-libs/aiohttp-session/ which we considered, however similar functionality could be achieved without using that library (which i think is implemented with the current version), well there is a need to go redis.

**Comment by blankdots** [comment]
with this AccessHandler methods `_save_to_session` and `_get_session` are obsolete do i understand it correctly ?

**Comment by pontus** [comment]
Correct, those are obsoleted (but were left as I thought they were still in use). Removed by 45a0112.

File: metadata_backend/api/auth.py [hunk 1]
```diff
@@ -87,9 +88,8 @@ async def callback(self, req: Request) -> Response:
             LOG.error(reason)
             raise web.HTTPBadRequest(reason=reason)
 
-        # Verify, that states match
-        state = await self._get_from_session(req, "oidc_state")
-        if not secrets.compare_digest(str(state), str(params["state"])):
+        # Verify, that state is pending
+        if not params["state"] in req.app["OIDC_State"]:
```
Comment (blankdots) [inline metadata_backend/api/auth.py:92]: wondering why we don't need to compare digest here or why we dropped it ?

File: metadata_backend/api/middlewares.py [hunk 2]
```diff
@@ -30,6 +30,7 @@ async def http_error_handler(req: Request, handler: Callable) -> Response:
     except web.HTTPError as error:
         details = _json_exception(error.status, error, req.url)
         LOG.error(details)
+        # LOG.error(error.__traceback__.format_exc())
```
Comment (blankdots) [inline metadata_backend/api/middlewares.py:33]: forgotten log ?

File: metadata_backend/api/auth.py (hunk 1 above)
Comment (pontus) [reply metadata_backend/api/auth.py:92]: With this PR, we delay session set up (cookie creation) to the callback, so we don't have the connection to compare the specific handed out `state` to the received. As a second best thing, we keep track of currently outstanding authentications (where callbacks have not yet been called).

I'm not sure I see understand what the security benefit of verifying the state is, but if one want to keep track of it, one can move the session creation (cookie setting) to `login`. The drawback of that (reason I did not do it for this PR) is that the cookie can not be completely filled in at that time and I've seen overall weird behaviour when changing cookies (cookies on redirects are also a thing, and I don't have a complete overview for what cases it works and when it's ignored).

File: metadata_backend/api/middlewares.py (hunk 2 above)
Comment (pontus) [reply metadata_backend/api/middlewares.py:33]: Yes, apologies.

**Review by blankdots — APPROVED** [review]
I think we can have this merged into develop. No objections.

**Pull request merged**